Website

davidrudlstorfer opened the pull request.

davidrudlstorfer commented:
As written in #197 the problem was that this does not work for forks, therefore, I've temporarily added the website branch to the main repo

TODO => remove my admin rights @isteinbrecher after merge of the new website

:partying_face: naturally it works and the new website is live on https://imcs-compsim.github.io/meshpy/

davidrudlstorfer commented:
@isteinbrecher the images will work once this PR is merged and the images are available on the main branch

davidrudlstorfer commented:
Last point from my side: In the future it is also mandatory that the website building and deployment only runs once the test suite has completed. For that I've added

```
  # Only build and deploy the website once the testsuite is completed => coverage badge and report are necessary
  workflow_run:
    workflows: ["Test suite"]
    types:
      - completed
```

This only works on the main branch -> therefore, we have to wait until it is merged if it works

isteinbrecher commented on README.md:
In the line
```
Feel free to leave a :star: on [GitHub](https://github.com/imcs-compsim/meshpy).
```
the `:star:` is displayed wrong on the website. If there is no quick fix, I would just change it to `star`.

davidrudlstorfer replied:
The notation for emojis which uses `:emoji:` is only used by Github. Now the readme is rendered with the myst parser, therefore, we need to exchange/add emojis with their unicode counterpart (on macOS we can directly add the emojis via the keyboard)

I thought that I do not update too much in the readme in this PR but I've exchanged the emoji

isteinbrecher reviewed (commented):
Thanks for the changes @davidrudlstorfer looks great! (not just compared to the old website).

I left some comments.

What is the default workflow when changing something in the website? Can one build the website locally before committing? I think we should (very briefly) document this somewhere.

I really like the prominent GitHub link, I think it is really annoying if one can't easily go to GitHub from the website.

davidrudlstorfer commented on .github/workflows/github-pages.yml:
```diff
@@ -1,63 +1,66 @@
+name: Build and deploy website
 
 on:
+  schedule:
```
Good point - after giving it a second thought it is unnecessary. Initially I thought that it is necessary to constantly update the website (in the future it will host the coverage badge/report) to account for changes in the coverage. But it is rebuilt anyways after each merge to main. I removed it

davidrudlstorfer commented on website/docs/source/conf.py:
```diff
@@ -0,0 +1,77 @@
+# https://www.sphinx-doc.org/en/master/usage/configuration.html#project-information
+
+project = "MeshPy"
+copyright = "2025, MeshPy"
+author = "MeshPy"
```
Changed it accordingly

davidrudlstorfer commented on README.md:
```diff
@@ -55,7 +55,7 @@ A few things to keep in mind:
 - MeshPy uses `black` to format python code.
   Make sure to apply `black` to the changed source files.
-- Feel free to add yourself to the [CONTRIBUTORS](CONTRIBUTORS) file.
+- Feel free to add yourself to the [CONTRIBUTORS](https://github.com/imcs-compsim/meshpy/blob/main/CONTRIBUTORS) file.
```
I'll explain it further down below.

davidrudlstorfer commented on README.md:
```diff
@@ -155,3 +155,20 @@ If the ArborX extension is working correctly can be checked by running the geome
+## Examples
+
+<p align="center">
+  <img src="https://github.com/imcs-compsim/meshpy/utilities/assets/images_readme/composite_plate.png" width="350" title="Fiber reinforced composite plate">
```
I'll explain it further down below.

davidrudlstorfer commented on utilities/assets/images_readme/composite_plate.png:
I'll explain it further down below.

davidrudlstorfer commented:
@isteinbrecher thanks for your review!

It's a good idea to document the website building process somewhere and also how to add changes. I've readded a `README.md` in the website directory and added the necessary steps there.

To only change contents on the main landing page no adjustments within the website directory are necessary. The top level readme is directly linked. Therefore, simply adjust the `README.md` and the website automatically reflects these changes.

Now regarding the absolute links for other files and images:

The underlying idea of the new website is to directly "copy" or reflect the `README.md` of MeshPy to decrease the necessary workload for the website. This approach can be achieved via 2 ways:

1. Create a new second website file which needs to be manually updated everytime something changes in the `README.md`
2. Automatically reflect the `README.md` on the website.

I think the second approach is nicer because the website won't be outdated if someone forgets to update the website.

I achieved this approach by symlinking the `README.md` into the website directory (`source/index.md`). With this the contents always stay up to date.

And this brings me to the necessary changes regarding links and images within the `README.md`. We now want to visualize the `README.md` in two ways - 1. locally when opening it and 2. on the website.

Therefore, links (to other files) and images need to be saved somewhere so that the website can access it and we can access it locally.

If the files are linked relatively, the website is not able to crosslink them. By providing the absolute Github link of the files and images we can view the `README.md` in both ways.

I added the images to the `utitlites/assets` because in my opinion they do not strictly belong to the website because we mainly use them in the `README.md`

And one final note regarding the `CONTRIBUTORS` file - I would remove it. As far as I know Github does not support that file and the contributors are listed on Github anyways. I would then only add the Contributors in the readme - this way we have them in the repo and on the website as well without duplicating information (which can be outdated again)

Looking forward to your opinion on this approach!

isteinbrecher commented on .github/workflows/website.yml:
```diff
@@ -0,0 +1,64 @@
+      - name: Install dependencies
+        run: |
+          python -m pip install --upgrade pip
+          pip install sphinx pydata-sphinx-theme myst-parser linkify-it-py
```
We could use the `website/requirements.txt` file here (the approach is simply copied from splinepy).

davidrudlstorfer replied:
Good idea - i totally forgot to check the workflow within splinepy - that would have saved me a lot of trouble 🤦‍♂️

Added in latest commit

isteinbrecher reviewed (commented):
> It's a good idea to document the website building process somewhere and also how to add changes. I've readded a `README.md` in the website directory and added the necessary steps there.

Perfect, that is exactly what I was looking for. Maybe we can add a `requirements.txt` file inside the website folder to keep track of all required packages for the website and install them from there in the `website` workflow.

> To only change contents on the main landing page no adjustments within the website directory are necessary. The top level readme is directly linked. Therefore, simply adjust the `README.md` and the website automatically reflects these changes.
> ...
> If the files are linked relatively, the website is not able to crosslink them. By providing the absolute Github link of the files and images we can view the `README.md` in both ways.

For now, I am fine with this solution. In the future we can think of something similar to splinepy where they "replace" those links in the website workflow. This also eliminates the need for the symlink. I think this seems like a nice and clean approach.  

> I added the images to the `utitlites/assets` because in my opinion they do not strictly belong to the website because we mainly use them in the `README.md`

Can we move them to `utilities/doc`?

> And one final note regarding the `CONTRIBUTORS` file - I would remove it. As far as I know Github does not support that file and the contributors are listed on Github anyways. I would then only add the Contributors in the readme - this way we have them in the repo and on the website as well without duplicating information (which can be outdated again)

Good point, I took care of that already and added the commit to the PR.


All in all, I am really happy with this. I would say, let's add the last small fixes and then merge this. If we discover in the future, that some current solutions are not practicable, we can tackle them then.

isteinbrecher commented on README.md:
```diff
@@ -1,6 +1,7 @@
-</h1>
+<center>
+
+# MeshPy
```
Very nice, I already noticed that yesterday but forgot to comment.

davidrudlstorfer commented:
Thanks for your feedback - the URL exchange actually sounds like a good improvement for the future.

I've added the requirements file and also moved the assets into `utilities/doc` - let's add the changes regarding the contributors and then merge 🚀

isteinbrecher commented:
Ah, I uploaded to the wrong fork. Now the contributing should be resolved. Could you remove the "build website on push" and squash the commits? Then I think we can merge this!

davidrudlstorfer commented:
@isteinbrecher this would be ready

Once merged I'll remove the outdated stuff in the settings of the repo - then you could remove my admin rights again

isteinbrecher commented on README.md:
```diff
@@ -155,3 +156,32 @@ If the ArborX extension is working correctly can be checked by running the geome
+## Examples
+
+<p align="center">
+  <img src="https://github.com/imcs-compsim/meshpy/utilities/doc/composite_plate.png" width="350" title="Fiber reinforced composite plate">
```
I think we have to add the `tree/main` part in the URL here.

davidrudlstorfer replied:
I'm currently at it.

Somehow the header is not centered on Github, and the action runs although test suite is not finished

I'll open a new PR

isteinbrecher replied:
In the website the header looks fine. Maybe this is a GitHub display issue.

isteinbrecher reviewed (commented):
Great that this next chapter is modernised!

I think we need a small follow up PR fixing the links 😅

isteinbrecher commented:
I already deleted the website branch in the main repo, but it should be fine doing this with a branch in the fork.
